[Feature] Add support partitions in policy data sources

ashenm opened the pull request:
## Changes
- Resolves https://github.com/databricks/terraform-provider-databricks/issues/4054
- Resolves https://github.com/databricks/terraform-provider-databricks/issues/4152
- Add optional argument `aws_partition` to all aws policy data sources to allow usage in all aws partitions

## Tests
<!-- 
How is this tested? Please see the checklist below and also describe any other relevant tests 
-->

- [x] `make test` run locally
- [x] relevant change in `docs/` folder
- [x] covered with integration tests in `internal/acceptance`
- [x] relevant acceptance tests are passing
- [ ] using Go SDK

alexott commented on aws/data_aws_bucket_policy.go:
```diff
@@ -16,6 +16,7 @@ func DataAwsBucketPolicy() common.Resource {
 	return common.Resource{
 		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
 			bucket := d.Get("bucket").(string)
+			partition := d.Get("aws_partition").(string)
```
can we use the same naming everywhere, like `aws_partition`?

alexott commented on aws/data_aws_unity_catalog_policy.go:
```diff
@@ -92,6 +93,12 @@ func validateSchema() map[string]*schema.Schema {
+		"aws_partition": {
+			Type:         schema.TypeString,
+			Optional:     true,
+			ValidateFunc: validation.StringInSlice([]string{"aws", "aws-cn", "aws-us-gov"}, false),
```
can we define this slice as constant or var, so it will be accessible to all data sources?

ashenm replied:
Good call! Moved as a const to common

alexott commented on docs/data-sources/aws_bucket_policy.md:
```diff
@@ -75,6 +75,7 @@ resource "aws_s3_bucket_policy" "ds" {
 ## Argument Reference
 
 * `bucket` - (Required) AWS S3 Bucket name for which to generate the policy document.
+* `aws_partition` - (Optional) AWS partition. The options are `aws`, `aws-cn`, or `aws-us-gov`.
```
add that `aws` is the default value for this option? And do it in all documents?

ashenm replied:
Done 🙌

alexott reviewed (commented):
overall looks good, thank you - just small nits.

@nkvuong wdyt?

alexott commented on aws/data_aws_unity_catalog_assume_role_policy.go:
```diff
@@ -14,13 +15,19 @@ func DataAwsUnityCatalogAssumeRolePolicy() common.Resource {
 		}
+
+		if !slices.Contains(common.AwsPartitions, data.AwsPartition) {
+			return fmt.Errorf("aws_partition must be either 'aws', 'aws-cn', or 'aws-us-gov'")
```
nit: we can move error message into constants...

ashenm replied:
Done 🙌

alexott commented on common/aws.go:
```diff
@@ -0,0 +1,3 @@
+package common
+
+var AwsPartitions = []string{"aws", "aws-cn", "aws-us-gov"}
```
it's not necessary to be in the `common` - just create a separate file with constants/variables inside `aws` directory`

ashenm replied:
Moved to `aws/constants.go`

alexott commented on aws/data_aws_unity_catalog_policy.go:
```diff
@@ -15,6 +15,7 @@ import (
 func generateReadContext(ctx context.Context, d *schema.ResourceData, m *common.DatabricksClient) error {
 	bucket := d.Get("bucket_name").(string)
 	awsAccountId := d.Get("aws_account_id").(string)
+	awsPartition := d.Get("aws_partition").(string)
```
we need to add validation for a value here as well...

ashenm replied:
https://github.com/databricks/terraform-provider-databricks/blob/0c45e23268ddfd367934fea888f604828e6b2f23/aws/data_aws_unity_catalog_policy.go#L99 takes care of it or do we still need an extra validation here?

alexott replied:
ah, yes - missed it.

alexott reviewed (commented):
looks mostly good, just some comments

ashenm commented:
@hectorcast-db please take a stab at this

alexott reviewed (changes requested):
@ashenm I did look a bit deeper, and here are additional comments that we need to address:

* https://github.com/databricks/terraform-provider-databricks/blob/main/aws/data_aws_assume_role_policy.go isn't modified, especially, we need to double-check this one: https://github.com/databricks/terraform-provider-databricks/blob/main/aws/data_aws_assume_role_policy.go#L54 - the log delivery role is hardcoded
* for aws bucket policy, first, `databricks_account_id` isn't documented at all, and we need to put a comment there saying that it should be changed for non-aws partition

ashenm commented:
@alexott 

> https://github.com/databricks/terraform-provider-databricks/blob/main/aws/data_aws_assume_role_policy.go isn't modified, especially, we need to double-check this one: https://github.com/databricks/terraform-provider-databricks/blob/main/aws/data_aws_assume_role_policy.go#L54 - the log delivery role is hardcoded

Yah nice cath! Updated all log delivery, unity catalogue, account id references https://github.com/databricks/terraform-provider-databricks/compare/5ec45228a6784ff4b1e1402b6d6ab4d2938a8a84..dc244d4375e309f9d14963be4785619c90ff0df1

> for aws bucket policy, first, databricks_account_id isn't documented at all, and we need to put a comment there saying that it should be changed for non-aws partition

Yep not just bucket policy assume role policy too don't have `databricks_account_id` listed as a parameter; added depreciation message on both and to resolev account id depending on partition being selected when not provided

ashenm commented:
@alexott any other codeowners apart from @nkvuong?

ashenm commented:
Thanks @nkvuong but looks like yours and @alexott's approvals aint enough 😬

ashenm commented:
Not sure about the branch protection rules maybe @mgyucht @tammyma-db @nfx @hectorcast-db maybe one of y'all can additionally approve? (judging by past commits on these files :P)

ashenm commented:
Thanks @alexott one more help how to trigger the integration tests? It seems it's awaiting its succession

alexott commented:
it will be merged when tests finished. don't worry...

ashenm commented:
@alexott yeah but the test trigger has been skipped https://github.com/databricks/terraform-provider-databricks/actions/runs/11736462058/job/32699657127

ashenm commented:
@alexott thanks for triggering those worflows; https://go/deco-tests-run/terraform must be also triggered manually I beilieve since my user don't have access to org secrets like _DECO_WORKFLOW_TRIGGER_APP_ID_?

ashenm commented:
@alexott any chance to see what's failing on the integration tests?

ashenm commented:
Nvmd; bucket policy integrations have a overriding test bucket fill allow override
